### ISSUE TITLE
Salesforce randomly failing API requests ("Invalid JSON response from Salesforce")

### ISSUE THREAD

**Issue by githubbob42** [githubbob42/mingle2github2#2414]
**[Mingle Card: 2682](https://liquidframeworks.mingle.thoughtworks.com/projects/alpine_mobile/cards/2682)**
check with Yusuke before you work on this, he might be working on this already.

Salesforce randomly returns invalid JSON responses when making API calls. The error message can take 2 possible forms on the sync screen:

1) "Error: Invalid JSON response from Salesforce\\n at Object.errors.invalidJson (/Users/ymorita/Work/alpine-mobile/src/node\_modules/nforce/index.js:1014:12)\\n at Request.\_callback (/Users/ymorita/Work/alpine-mobile/src/node\_modules/nforce/index.js:1131:34)\\n at Request.self.callback (/Users/ymorita/Work/alpine-mobile/src/node\_modules/request/request.js:121:22)\\n at Request.EventEmitter.emit (events.js:98:17)\\n at Request.<anonymous> (/Users/ymorita/Work/alpine-mobile/src/node\_modules/request/request.js:985:14)\\n at Request.EventEmitter.emit (events.js:117:20)\\n at IncomingMessage.<anonymous> (/Users/ymorita/Work/alpine-mobile/src/node\_modules/request/request.js:936:12)\\n at IncomingMessage.EventEmitter.emit (events.js:117:20)\\n at \_stream\_readable.js:920:16\\n at process.\_tickCallback (node.js:415:13)\\n"

2) "Invalid JSON response from Salesforce"

Cause is unknown, but the failing response usually looks like a partial JSON result and the following appended to the response body (see attached samples):

```
\[{"message":"An unexpected error occurred.","errorCode":"UNKNOWN\_EXCEPTION"}\]
```

This implies Salesforce has no idea what the error is either...

I've seen this error while fetching contacts and pricebooks which implies it's not tied to any particular client code path, but rather the API being used: 

```
https://na15.salesforce.com/services/data/v28.0/query
```

Tabling this for now since it happens infrequently (1~2 times per day).

WIP can be found here: https://github.com/pixel9/alpine-mobile/compare/invalid-json

Addtional Dev Notes
-------------------

These errors recieved back from salesforce or the nforce module. 

The error An unexpected error occurred. is recieved from salesforce.

The error Error: Invalid JSON response from Salesforce... is part nforce library part salesforce. The nforce module prefixes the errror with "Error: Invalid JSON response from Salesforce" followed by the rest of the salesforce error. 

Test Plan
---------

**Comment by cassiodias** [comment]
Any update regarding this "Invalid JSON response from Salesforce"?

**Comment by githubbob42** [comment]
This is a just test repo I used to test the migration our cards/tickets from ThoughtWorks/Mingle to Github before importing into our real repo, so everything in this repo is just a snapshot in time and not used or maintained.

However, the ticket in question is VERY old and we already closed and deleted it from our real repo.  Like the card said, it may have been due to the API version being used but it was most likely to have been to due to using `window.fetch` which does not (or did not) handle chunked responses, so naturally, any chunked response that isn't handled will fail when trying to parse the returned JSON.

When we upgraded our sync code, our architect at the time started using `window.fetch` because it was new and shiny.  We ended up patching `nforce` to handle the chunked responses.  It wasn't until later (after he left) that we found about the deficiencies in `fetch` and rewrote our sync engine without it and haven't had any issues since.
